Fix issue #20

charleswhchan opened the pull request:
README.rst
* Added step to start serf agent before running tests.

serfclient/connection.py
* Added custom data unpacker for `Addr` field in `members` response message.

charleswhchan commented:
(PR is a WIP. Currently, it returns `'Addr': '::ffff:192.168.1.108'`)

@KushalP Before completing the rest of the changes, I would like to consult and see if you have any preference on the data format ...

1. IPv4 address format:
a) `::ffff:192.168.1.108` - ie. keep as is
b) `192.168.1.108`- ie. dotted-quad string format (my preference)
Note: IPv6 address will return the long hex format.

2. brackets & comma?
Recall in the Serf document (https://www.serfdom.io/docs/agent/rpc.html), the IP is returned as `"Addr": [127, 0, 0, 1]`. Do you want the dictionary value to follow the same format? 
(my preference: return as a string directly `'Addr': '127.0.0.1'`)

Thoughts?

KushalP commented:
This looks like a great start, and I'm looking forward to the tests you mention so that we don't have regressions :smile:

As for your questions:
1. If we receive an IPv4 address we should return a dotted-quad string (can you add a test for the IPv6 form as well please?)
2. Returning as a string is fine

KushalP commented on serfclient/connection.py:
```diff
@@ -134,3 +134,14 @@ def _counter(self):
+    def _decode_object(self, obj_dict):
+        # Workaround for issue 20:
+        # Serf packs 'Addr' as an IP address, and can be unpack using
+        # socket.inet_ntop.
```
I think these comments should either be removed or made more useful to those reading them. If we're going to keep them, we should make them [docstrings](https://www.python.org/dev/peps/pep-0257/#id17) and provide a link directly to the issue in question.

charleswhchan replied:
Alright. I will turn them into docstrings.

KushalP commented on serfclient/connection.py:
```diff
@@ -134,3 +134,14 @@ def _counter(self):
         return "Error %s connecting %s:%s. %s." % \
             (exception.args[0], self.host, self.port, exception.args[1])
+
+    def _decode_object(self, obj_dict):
```
Can you provide a more descriptive method name here? e.g. `_decode_addr_key`

charleswhchan replied:
Sure. (Believe it or not, I actually used that name initially...) :smile:

KushalP replied:
:ok_hand:

charleswhchan commented:
>  (can you add a test for the IPv6 form as well please?)

I was looking at test_client.py earlier, and the output format IPv4/v6 may be tricky to control since it is dependent on local inet config and serf agent. Any suggestion or example I could refer to?

Thanks.

KushalP commented:
> the output format IPv4/v6 may be tricky to control since it is dependent on local inet config and serf agent. Any suggestion or example I could refer to?

That's a fair point. Building out that functionality could be quite tricky.

To keep this PR momemtum up, let's concentrate on decoding the field for now. We can look at the full integration-level testing for IPv6 later. However, if you come up with a suitable refactoring that works, I'm all ears :smile:

charleswhchan commented:
Updated code based on review comments:
* Converted comments to docstrings.
* Format IPv4 mapped IPv6 to dotted-quad notation.
* Add unit test to verify IPv4 format using regexp.

Trying to find a way to test IPv6. Tried to do the following but failed:
```
./serf agent --tag foo=bar --advertise=2001:db8:85a3:0:0:8a2e:370:7334
==> Invalid advertise address: too many colons in address 2001:db8:85a3:0:0:8a2e:370:7334
```
Not sure if this is a bug in Serf? I will play around with it ...

charleswhchan commented:
Hum. Looks like the tests are failing in Python 3.x (I am using 2.7), any ideas?

charleswhchan commented:
@KushalP I have fixed the Python 3.x failures. Please review.

I have tested the code in both Python 2.7 and 3.4 to make sure it returns and IPv4 quad dotted address: `'Addr': '192.168.0.1'`.

I also ran a test manually with an IPv6 address `./serf agent -bind=[::1]`; and I get back an `'Addr': '::1'`.

Finaly, if you are satisfied with the changes, could you please generate a new release on PyPi? 

Many thanks!

KushalP commented:
Thanks :+1:

KushalP commented:
I've just released [v0.6.0](https://pypi.python.org/pypi/serfclient/0.6.0).

charleswhchan commented:
> I've just release v0.6.0.

Many thanks, @KushalP!!

TechBK commented on serfclient/connection.py:
```diff
@@ -134,3 +134,28 @@ def _counter(self):
+        """
+        key = b'Addr'
+        if key in obj_dict:
+            ip_addr = socket.inet_ntop(socket.AF_INET6, obj_dict[key])
```
Here, code will occur error if `obj_dict[key]` is IPv4.
My suggest:
```python
        if key in obj_dict:
            try
                ip_addr = socket.inet_ntop(socket.AF_INET6, obj_dict[key])

                # Check if the address is an IPv4 mapped IPv6 address:
                # ie. ::ffff:xxx.xxx.xxx.xxx
                if ip_addr.startswith('::ffff:'):
                ip_addr = ip_addr.lstrip('::ffff:')

                obj_dict[key] = ip_addr.encode('utf-8')
        except:
            pass
        return obj_dict
```

charleswhchan replied:
Thanks @TechBK. You are right, I consulted the Python doc for [socket.inet_ntop()](https://docs.python.org/2/library/socket.html#socket.inet_ntop) and ran the following test. 

In summary, a ValueError could be raised. I will submit a PR to fix.
```
>>> import socket
>>> socket.inet_ntop(socket.AF_INET6, '192.168.0.1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid length of packed IP address string
```

TechBK replied:
What your python you used ? Python3 or python2?
I run line `socket.inet_ntop(socket.AF_INET6, '192.168.0.1')` on python3, but it raise error:
```bash
Traceback (most recent call last):
  File "/home/vcteam/.pycharm_helpers/pydev/pydevd.py", line 1531, in <module>
    globals = debugger.run(setup['file'], None, None, is_module)
  File "/home/xxxx/.pycharm_helpers/pydev/pydevd.py", line 938, in run
    pydev_imports.execfile(file, globals, locals)  # execute the script
  File "/home/xxxx/.pycharm_helpers/pydev/_pydev_imps/_pydev_execfile.py", line 18, in execfile
    exec(compile(contents+"\n", file, 'exec'), glob, loc)
  File "/home/xxxx/plugins/learn/learn_xxx.py", line 4, in <module>
    print(socket.inet_ntop(socket.AF_INET, '192.168.145.21'))
TypeError: 'str' does not support the buffer interface
```

charleswhchan replied:
@TechBK. Ignore the 'str' test.

```
$ python3
Python 3.4.3 (default, Oct 14 2015, 20:28:29) 
>>> import socket
>>> socket.inet_ntop(socket.AF_INET6, socket.inet_pton(socket.AF_INET, '192.168.0.1'))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: invalid length of packed IP address string
```

TechBK replied:
Show me why it passed ?
i dont know it :D
